File: 2025-10-07 - Padroes de Projetos - Iterator/Exemplo_Python/EstanteIterator.py
class EstanteIterator:
   def __init__(self, livros):
      self.__livros = livros
      self.__posicao = 0

   def getLivros(self):
      return self.__livros

   def setLivros(self, livros):
      self.__livros = livros

   def getPosicao(self):
      return self.__posicao

   def setPosicao(self, posicao):
      self.__posicao = posicao

   def __iter__(self):
      return self

   def __next__(self):
      if self.__posicao < len(self.__livros) and self.__livros[self.__posicao] is not None:
         livro = self.__livros[self.__posicao]
         self.__posicao += 1
         return livro
      else:
         raise StopIteration

File: 2025-10-07 - Padroes de Projetos - Iterator/Exemplo_Java/EstanteIterator.java
import java.util.Iterator;
import java.util.List;

public class EstanteIterator implements Iterator<Livro> {
   private List<Livro> livros;
   private int posicao = 0;

   public EstanteIterator(List<Livro> livros) {
      this.livros = livros;
   }

   public List<Livro> getLivros() {
      return livros;
   }

   public void setLivros(List<Livro> livros) {
      this.livros = livros;
   }

   public int getPosicao() {
      return posicao;
   }

   public void setPosicao(int posicao) {
      this.posicao = posicao;
   }

   @Override
   public boolean hasNext() {
      return posicao < livros.size() && livros.get(posicao) != null;
   }

   @Override
   public Livro next() {
      return livros.get(posicao++);
   }
}
